### ISSUE TITLE
The behaviour of FIFO queues does not correspond with AWS specifications

### ISSUE THREAD

**Issue by robmosca** [spulec/moto#2433]
## Reporting Bugs

The test code reported in issue #1655 does not correspond with the behaviour of AWS SQS and the specifications of boto3.

As the boto3 documentation says:

> When you set this attribute (FifoQueue), you **must also provide the MessageGroupId for your messages explicitly**. For more information, see FIFO Queue Logic in the Amazon Simple Queue Service Developer Guide.

This is currently not enforced in moto.

Furthermore, as the AWS says:

> When receiving messages from a FIFO queue with multiple message group IDs, Amazon SQS first attempts to return **as many messages with the same message group ID as possible**. This allows other consumers to process messages with a different message group ID.

Currently only one message for each message group ID is returned and the next one is not returned until the first one is deleted.

Currently:

1. It is possible to call send_message for a FIFO queue without specifying MessageGroupID
2. When calling `receive_messages()` on a queue with multiple messages with the same Message Group ID only one message is returned and the next one is not returned until the first one is deleted.

If I am not mistaken the specification should be something like this:

```python
@mock_sqs
def test_receive_messages_with_message_group_id():
    sqs = boto3.resource('sqs', region_name='us-east-1')
    queue = sqs.create_queue(QueueName="blah.fifo", Attributes={'FifoQueue': 'true', 'ContentBasedDeduplication': 'true'})
    queue.send_message(
        MessageBody="message-1",
        MessageGroupId="group"
    )
    queue.send_message(
        MessageBody="message-2",
        MessageGroupId="group"
    )

    messages = queue.receive_messages(MaxNumberOfMessages=10)
    messages.should.have.length_of(2)

@mock_sqs
def test_send_messages_to_fifo_without_message_group_id():
    sqs = boto3.resource('sqs', region_name='us-east-1')
    queue = sqs.create_queue(QueueName="blah.fifo", Attributes={'FifoQueue': 'true', 'ContentBasedDeduplication': 'true'})

    with self.assertRaises(Exception):
        queue.send_message(
            MessageBody="message-1"
        )
```

**Comment by spulec** [comment]
Thanks; going to set this as a feature request to implement the entire spec for this.

**Comment by bblommers** [comment]
This is now fixed in Moto >= 1.3.15.dev874
